packages/router/router.ts

```diff
@@ -3106,35 +3312,68 @@ async function callLoaderOrAction(
   let resultType;
   let result;
 
-  // Setup a promise we can race against so that abort signals short circuit
-  let reject: () => void;
-  let abortPromise = new Promise((_, r) => (reject = r));
-  let onReject = () => reject();
-  request.signal.addEventListener("abort", onReject);
+  let onReject: (() => void) | undefined;
 
   try {
-    let handler = match.route[type];
-    invariant<Function>(
-      handler,
-      `Could not find the ${type} to run on the "${match.route.id}" route`
-    );
+    // Load any lazy route modules as part of the loader/action phase
+    if (match.route.lazy) {
+      await loadLazyRouteModule(match.route, hasErrorBoundary, manifest);
+    } else {
+      invariant<Function>(
+        match.route[type],
+        `Could not find the ${type} to run on the "${match.route.id}" route`
+      );
+    }
 
-    result = await Promise.race([
-      handler({ request, params: match.params, context: requestContext }),
-      abortPromise,
-    ]);
+    let handler = match.route[type];
+    if (!handler) {
+      if (type === "action") {
+        throw getInternalRouterError(405, {
+          method: request.method,
+          pathname: new URL(request.url).pathname,
+          routeId: match.route.id,
+        });
+      } else {
+        // lazy() route has no loader to run
+        result = undefined;
+      }
+    } else {
+      // Setup a promise we can race against so that abort signals short circuit
+      let reject: () => void;
+      let abortPromise = new Promise((_, r) => (reject = r));
+      onReject = () => reject();
+      request.signal.addEventListener("abort", onReject);
+
+      if (!request.signal.aborted || type === "action") {
+        // Still kick off actions if we got interrupted to maintain consistency
+        // with un-abortable behavior of action execution on non-lazy routes
```

Comment:
@mjackson @markdalgleish Thinking about this - is there any reason not to let this apply to loaders as well?  Both for code simplicity as well as consistency.  I.e., you cannot prevent a loader from running once a route is defined.

Reply:
Make this consistent with static behavior for both loaders and actions